### PR TITLE
permit avails research over multi-term mandates

### PR DESCRIPTION
**Pull request by yspringfield**
Fixed: #7887

### CODE REVIEW

File: apps/catalog/catalog/src/app/marketplace/title/list/list.component.ts [hunk 1]
```diff
@@ -152,19 +152,24 @@ export class ListComponent implements OnDestroy, OnInit {
...
       this.snackbar.open('This title is not available', 'close', { duration: 5000 });
       return;
     }
 
-    this.bucketService.addTerm(title.objectID, parentTerm.id, this.availsForm.value);
+    const results = availResults.map(({ mandateOrTerm: term, avail }) => ({
+      titleId: title.objectID,
+      parentTermId: term.id,
+      avail
+    }));
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/marketplace/title/list/list.component.ts:170]:
```suggestion
    const results = availResults.map(res => ({
      titleId: res.title.objectID,
      parentTermId: res.manateOrTerm.id,
      avail: res.avail
    }));
```

File: libs/contract/src/lib/avails/avails.ts [hunk 2]
```diff
@@ -83,6 +88,112 @@ function assertValidTitle(mandates: FullMandate[], sales: FullSale[], bucketCont
...
+interface SearchAvailsConfig<A extends BaseAvailsFilter> {
+  avails: A,
+  mandates: FullMandate[],
+  sales?: FullSale[],
+  mandateFilterFn: (term: Term<Date>, avail: A) => boolean,
+  saleFilterFn?: (term: Term<Date>, avail: A) => boolean,
+  verifyDurationIntersection: boolean
+};
+
+function getMatchingAvailabilities<A extends AvailsFilter | CalendarAvailsFilter>(options: SearchAvailsConfig<A>) {
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/avails/avails.ts:100]: rename `options` to `config`

File: libs/contract/src/lib/avails/avails.ts [hunk 3]
```diff
@@ -83,6 +88,112 @@ function assertValidTitle(mandates: FullMandate[], sales: FullSale[], bucketCont
...
+
+  const results: AvailResult<A>[] = [];
+
+  const subAvails = avails.territories.map(ter => {
+    return avails.medias.map(m => ({
+      ...avails,
+      territories: [ter],
+      medias: [m]
+    }));
+  }).flat();
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/avails/avails.ts:118]:
```suggestion
  const subAvails = avails.territories.map(territory => {
    return avails.medias.map(media => ({
      ...avails,
      territories: [territory],
      medias: [media]
    }));
  }).flat();
```

File: libs/contract/src/lib/avails/avails.ts [hunk 4]
```diff
@@ -92,43 +202,51 @@ export interface AvailsFilter extends BaseAvailsFilter {
...
 
-    return exclusivityCheck && mediaCheck && durationCheck && territoryCheck;
-  }));
+export function getListMatchingAvailabilities(mandates: FullMandate[], sales: FullSale[], avails: AvailsFilter) {
+  const options = {
+    mandates, sales, avails,
+    mandateFilterFn: isAvailAllInTerm,
+    saleFilterFn: isListAvailPartiallyInTerm,
+  };
+  return getMatchingAvailabilities(options);
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/avails/avails.ts:231]:
```suggestion
  const config = {
    mandates,
    sales,
    avails,
    mandateFilterFn: isAvailAllInTerm,
    saleFilterFn: isListAvailPartiallyInTerm,
  };
  return getMatchingAvailabilities(config);
```

File: libs/contract/src/lib/avails/avails.ts [hunk 5]
```diff
@@ -92,43 +202,51 @@ export interface AvailsFilter extends BaseAvailsFilter {
...
+export function getMandateTerms(avails: AvailsFilter, mandates: FullMandate[]): AvailSearchResult<AvailsFilter>[] | undefined {
+  const options = {
+    mandates, avails,
+    mandateFilterFn: isAvailAllInTerm,
+  };
+  const { available } = getMatchingAvailabilities(options);
+  return available.map(a => ({
+    term: a.mandate.terms[0],
+    avail: a.avail
+  }));
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/avails/avails.ts:249]:
```suggestion
export function getMandateTerms(avails: AvailsFilter, mandates: FullMandate[]): AvailSearchResult<AvailsFilter>[] | undefined {
  const config = {
    mandates,
    avails,
    mandateFilterFn: isAvailAllInTerm,
  };
  const { available } = getMatchingAvailabilities(options);
  return available.map(a => ({
    term: a.mandate.terms[0],
    avail: a.avail
  }));
```

File: libs/contract/src/lib/avails/avails.ts [hunk 6]
```diff
@@ -422,85 +537,16 @@ export function isCalendarTermInAvails<T extends BucketTerm | Term>(term: T, ava
...
-    available,
-    sold
-  }
+export function getMatchingCalendarAvailabilities(avails: CalendarAvailsFilter, mandates: FullMandate[], sales: FullSale[]) {
+  const options = {
+    avails, mandates, sales,
+    mandateFilterFn: isCalendarTermInAvails,
+    saleFilterFn: isCalendarAvailPartiallyInTerm,
+  };
+  return getMatchingAvailabilities(options);
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/avails/avails.ts:546]:
```suggestion
  const config = {
    avails,
    mandates,
    sales,
    mandateFilterFn: isCalendarTermInAvails,
    saleFilterFn: isCalendarAvailPartiallyInTerm,
  };
  return getMatchingAvailabilities(config);
```

File: libs/contract/src/lib/bucket/+state/bucket.service.ts [hunk 7]
```diff
@@ -119,28 +125,22 @@ export class BucketService extends CollectionService<BucketState> {
...
+      const term = createBucketTerm(avail);
+      const sale = contracts.find(c => c.titleId === titleId && c.parentTermId === parentTermId);
+      if (sale) {
+        sale.terms.push(term);
+      } else {
+        const contract = createBucketContract({ titleId, parentTermId, terms: [term] });
+        contracts.push(contract);
+      }
     }
+    bucket
```
Comment (GrandSchtroumpf) [inline libs/contract/src/lib/bucket/+state/bucket.service.ts:142]: You should return here to be able to do `await` in front of call.
```suggestion
    return bucket
```

Comment (yspringfield) [reply libs/contract/src/lib/bucket/+state/bucket.service.ts]: it works well without an await, but I added it in for sake of doubt.

**Pull request merged**